### PR TITLE
Add SPMI benchmarks run collections for tiered and tiered pgo

### LINKED ISSUE #68179: SuperPMI benchmarks run should enable tiered compilation and pgo
Currently we run the benchmarks SPMI collection with TC disabled (and R2R disabled), and just run a single iteration. Among other things this means we're not collecting any OSR instances.

This differs from the product default that has TC enabled and now OSR. It would be nice to find a way to collect data that more closely represents how these benchmarks are actually run in the lab and by customers. This likely requires allowing BDN to go something closer to its normal iteration strategy, which in turn may end up taking a much longer time (eg a full run takes ~16+ hours on one machine; I don't know how many partitions we create). I think we can tamp down the final iteration count and just run one so perhaps it will not be too bad.

As a bonus it would also be nice to enable a similar collection with dynamic PGO.

There is a small risk that SPMI + checked jit (if we use one) will create altered timing and so the collection still won't reflect the actual usage (eg methods may tier up more slowly and so impact BDN's iteration strategy).

category:eng-sys
theme:super-pmi
skill-level:intermediate
cost:medium
impact:medium

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Add two new run configurations for SPMI benchmarks: tiered and tiered pgo. So benchmark runs now have 3 separate collections.

The new ones are named "run_tiered" and "run_pgo", eg

```
benchmarks.run.windows.x64.checked.mch
benchmarks.run_tiered.windows.x64.checked.mch
benchmarks.run_pgo.windows.x64.checked.mch
```

Fixes #68179.

### CODE REVIEW

**Comment by EgorBo** [comment]
Is there a list of benchmarks we run to get a collection? afair the full run of them takes 8 hours or so

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

Ran this on the internal pipeline with temp guid `16ec4bcb-fe04-4c0b-9a72-617e2a2a9aee`. Here is a summary of results for one set of three:

```
;; "classic" collection with tiering disabled
;; mcs.exe -jitflags benchmarks.run.windows.x64.checked.mch

Grouped Flag Appearances (35507 contexts)

bits                count  percent  parsed
0000000048800000     5168   14.55%  SKIP_VERIFICATION IL_STUB BBOPT
0000000248800000       11    0.03%  SKIP_VERIFICATION IL_STUB BBOPT PUBLISH_SECRET_PARAM
0000000040800010    30033   84.58%  DEBUG_INFO SKIP_VERIFICATION BBOPT
0000002040800010        4    0.01%  DEBUG_INFO SKIP_VERIFICATION BBOPT REVERSE_PINVOKE
00000000c0800010      287    0.81%  DEBUG_INFO SKIP_VERIFICATION BBOPT FRAMED
0000000000800030        1    0.00%  DEBUG_INFO MIN_OPT SKIP_VERIFICATION
0000000080800030        3    0.01%  DEBUG_INFO MIN_OPT SKIP_VERIFICATION FRAMED

Individual Flag Appearances

   30328   85.41%  DEBUG_INFO
       4    0.01%  MIN_OPT
   35507  100.00%  SKIP_VERIFICATION
    5179   14.59%  IL_STUB
   35503   99.99%  BBOPT
     290    0.82%  FRAMED
      11    0.03%  PUBLISH_SECRET_PARAM
       4    0.01%  REVERSE_PINVOKE

;; "current default" collection with tiering enabled
;; mcs.exe -jitflags benchmarks.run_tiered.windows.x64.checked.mch

Grouped Flag Appearances (57701 contexts)

bits                count  percent  parsed
0000000048800000     5720    9.91%  SKIP_VERIFICATION IL_STUB BBOPT
0000000248800000        7    0.01%  SKIP_VERIFICATION IL_STUB BBOPT PUBLISH_SECRET_PARAM
0000008000800010    34989   60.64%  DEBUG_INFO SKIP_VERIFICATION TIER0
0000000040800010     8608   14.92%  DEBUG_INFO SKIP_VERIFICATION BBOPT
0000010040800010     7521   13.03%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1
0000002040800010        4    0.01%  DEBUG_INFO SKIP_VERIFICATION BBOPT REVERSE_PINVOKE
0000008080800010      288    0.50%  DEBUG_INFO SKIP_VERIFICATION FRAMED TIER0
00000100c0800010       36    0.06%  DEBUG_INFO SKIP_VERIFICATION BBOPT FRAMED TIER1
0000010040802010      510    0.88%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT TIER1
00000100c0802010       13    0.02%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT FRAMED TIER1
0000000000800030        2    0.00%  DEBUG_INFO MIN_OPT SKIP_VERIFICATION
0000000080800030        3    0.01%  DEBUG_INFO MIN_OPT SKIP_VERIFICATION FRAMED

Individual Flag Appearances

   51974   90.07%  DEBUG_INFO
       5    0.01%  MIN_OPT
     523    0.91%  OSR
   57701  100.00%  SKIP_VERIFICATION
    5727    9.93%  IL_STUB
   22419   38.85%  BBOPT
     340    0.59%  FRAMED
       7    0.01%  PUBLISH_SECRET_PARAM
       4    0.01%  REVERSE_PINVOKE
   35277   61.14%  TIER0
    8080   14.00%  TIER1

;; "pgo" collection with tiering and tieredPgo enabled
;; mcs.exe -jitflags benchmarks.run_pgo.windows.x64.checked.mch

Grouped Flag Appearances (88881 contexts)

bits                count  percent  parsed
0000000048800000     5842    6.57%  SKIP_VERIFICATION IL_STUB BBOPT
0000000248800000       14    0.02%  SKIP_VERIFICATION IL_STUB BBOPT PUBLISH_SECRET_PARAM
0000008100800010    34987   39.36%  DEBUG_INFO SKIP_VERIFICATION BBINSTR_IF_LOOPS TIER0
0000008020800010    11041   12.42%  DEBUG_INFO SKIP_VERIFICATION BBINSTR TIER0
0000000040800010     8608    9.68%  DEBUG_INFO SKIP_VERIFICATION BBOPT
0000010040800010     2865    3.22%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1
8600010040800010        8    0.01%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_METHOD_PROFILE HAS_DYNAMIC_PROFILE
a400010040800010      240    0.27%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_CLASS_PROFILE HAS_DYNAMIC_PROFILE
c400010040800010    16514   18.58%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_DYNAMIC_PROFILE
c600010040800010      335    0.38%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_METHOD_PROFILE HAS_DYNAMIC_PROFILE
e400010040800010     6643    7.47%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_CLASS_PROFILE HAS_DYNAMIC_PROFILE
e600010040800010       28    0.03%  DEBUG_INFO SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_CLASS_PROFILE HAS_METHOD_PROFILE HAS_DYNAMIC_PROFILE
0000002040800010        4    0.00%  DEBUG_INFO SKIP_VERIFICATION BBOPT REVERSE_PINVOKE
0000008180800010      287    0.32%  DEBUG_INFO SKIP_VERIFICATION FRAMED BBINSTR_IF_LOOPS TIER0
00000080a0800010       41    0.05%  DEBUG_INFO SKIP_VERIFICATION BBINSTR FRAMED TIER0
00000100c0800010       22    0.02%  DEBUG_INFO SKIP_VERIFICATION BBOPT FRAMED TIER1
c4000100c0800010      203    0.23%  DEBUG_INFO SKIP_VERIFICATION BBOPT FRAMED TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_DYNAMIC_PROFILE
e4000100c0800010        5    0.01%  DEBUG_INFO SKIP_VERIFICATION BBOPT FRAMED TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_CLASS_PROFILE HAS_DYNAMIC_PROFILE
c400010040802010      783    0.88%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_DYNAMIC_PROFILE
c600010040802010       74    0.08%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_METHOD_PROFILE HAS_DYNAMIC_PROFILE
e400010040802010      285    0.32%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_CLASS_PROFILE HAS_DYNAMIC_PROFILE
e600010040802010       33    0.04%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_CLASS_PROFILE HAS_METHOD_PROFILE HAS_DYNAMIC_PROFILE
c4000100c0802010       13    0.01%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT FRAMED TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_DYNAMIC_PROFILE
e4000100c0802010        1    0.00%  DEBUG_INFO OSR SKIP_VERIFICATION BBOPT FRAMED TIER1 HAS_PGO HAS_EDGE_PROFILE HAS_CLASS_PROFILE HAS_DYNAMIC_PROFILE
0000000000800030        2    0.00%  DEBUG_INFO MIN_OPT SKIP_VERIFICATION
0000000080800030        3    0.00%  DEBUG_INFO MIN_OPT SKIP_VERIFICATION FRAMED

Individual Flag Appearances

   83025   93.41%  DEBUG_INFO
       5    0.01%  MIN_OPT
    1189    1.34%  OSR
   88881  100.00%  SKIP_VERIFICATION
    5856    6.59%  IL_STUB
   11082   12.47%  BBINSTR
   42520   47.84%  BBOPT
     575    0.65%  FRAMED
   35274   39.69%  BBINSTR_IF_LOOPS
      14    0.02%  PUBLISH_SECRET_PARAM
       4    0.00%  REVERSE_PINVOKE
   46356   52.16%  TIER0
   28052   31.56%  TIER1
     478    0.54%  HAS_METHOD_PROFILE
   25165   28.31%  HAS_DYNAMIC_PROFILE
    7235    8.14%  HAS_CLASS_PROFILE
   24917   28.03%  HAS_EDGE_PROFILE
   25165   28.31%  HAS_PGO
```

**Comment by AndyAyersMS** [comment]
> Is there a list of benchmarks we run to get a collection? afair the full run of them takes 8 hours or so

The benchmarks get split into 30 sub pieces and farmed out to helix, so I think we may run them all?

**Comment by AndyAyersMS** [comment]
Also we are still collecting benchmarks runs with R2R disabled -- wonder if we should revisit that?

File: src/coreclr/scripts/superpmi.py [hunk 1]
```diff
@@ -295,6 +295,7 @@ def add_core_root_arguments(parser, build_type_default, build_type_help):
 collect_parser.add_argument("-mch_files", metavar="MCH_FILE", nargs='+', help="Pass a sequence of MCH files which will be merged. Required by --merge_mch_files.")
 collect_parser.add_argument("--use_zapdisable", action="store_true", help="Sets DOTNET_ZapDisable=1 and DOTNET_ReadyToRun=0 when doing collection to cause NGEN/ReadyToRun images to not be used, and thus causes JIT compilation and SuperPMI collection of these methods.")
 collect_parser.add_argument("--tiered_compilation", action="store_true", help="Sets DOTNET_TieredCompilation=1 when doing collections.")
+collect_parser.add_argument("--tiered_pgo", action="store_true", help="Sets DOTNET_TieredCompilation=1 and DOTNET_TieredPGO=1 when doing collections.")
```
Comment (BruceForstall) [inline src/coreclr/scripts/superpmi.py:298]: Maybe you should add a check post-parsing to error if both `tiered_compilation` and `tiered_pgo` are set.

**Comment by AndyAyersMS** [comment]
I was wondering why with tiering enabled we see a reasonably high fraction of methods bypassing tiering. Not we don't have an explicit flag for this so you have to deduce it by the absence of other flags. One such bin is
```
0000000040800010     8608    9.68%  DEBUG_INFO SKIP_VERIFICATION BBOPT
```
@EgorBo reminded me that dynamic methods (like the classic regex) are still not eligible for tiered compilation. 
See #73594

Sure enough regex seems to be a major contributor to this bin.

File: src/coreclr/scripts/superpmi_benchmarks.py [hunk 2]
```diff
@@ -89,6 +90,16 @@ def setup_args(args):
...
+    coreclr_args.verify(args,
+                        "tiered_compilation",
+                        lambda unused: True,
+                        "Unable to set tiered_compilation")
+
+    coreclr_args.verify(args,
+                        "tiered_pgo",
+                        lambda unused: True,
+                        "Unable to set tiered_pgo")
+
```
Comment (BruceForstall) [inline src/coreclr/scripts/superpmi_benchmarks.py:102]: Same here: error if both are used

File: src/coreclr/scripts/superpmi_collect_setup.py [hunk 3]
```diff
@@ -60,7 +60,7 @@
 
 is_windows = platform.system() == "Windows"
 
-legal_collection_types = [ "crossgen2", "pmi", "run" ]
+legal_collection_types = [ "crossgen2", "pmi", "run", "run_tiered", "run_pgo" ]
```
Comment (BruceForstall) [inline src/coreclr/scripts/superpmi_collect_setup.py:63]: Update help for `collection_type`, above

**Review by BruceForstall — CHANGES_REQUESTED** [review]
Just a few requests.

Actually, I'm happy to see not many changes were required.

**Comment by BruceForstall** [comment]
superpmi_benchmarks.py includes:

```
collection_command = f"{dotnet_exe} {benchmarks_dll}  --filter \"*\" --corerun {os.path.join(core_root, corerun_exe)} --partition-count {partition_count} " \
                         f"--partition-index {partition_index} --envVars DOTNET_JitName:{shim_name} " \
                         " DOTNET_ZapDisable:1  DOTNET_ReadyToRun:0 " \
                         "--iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 --strategy ColdStart --logBuildOutput"
```

So, R2R is disabled.

Do the other BDN arguments make sense for these new collections?

**Comment by AndyAyersMS** [comment]
> Do the other BDN arguments make sense for these new collections?

Hmm, good point. Let me see if we can afford to run the benchmarks "normally" or if it takes too long. I may leave things as is for nontiered as the codegen is not timing sensitive.

We are also using a checked runtime which tiers up aggressively and uses the wrong corelib (#60947). So these collections are not as representative of what actually happens as I'd like.

**Comment by kunalspathak** [comment]
> > Is there a list of benchmarks we run to get a collection? afair the full run of them takes 8 hours or so
> 
> The benchmarks get split into 30 sub pieces and farmed out to helix, so I think we may run them all?

Yes, we run them all with `-iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 ` so they complete quickly.

**Comment by EgorBo** [comment]
> > > Is there a list of benchmarks we run to get a collection? afair the full run of them takes 8 hours or so
> > 
> > 
> > The benchmarks get split into 30 sub pieces and farmed out to helix, so I think we may run them all?
> 
> Yes, we run them all with `-iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 ` so they complete quickly.

Meaning that they don't tier up properly?

**Comment by AndyAyersMS** [comment]
> > > > Is there a list of benchmarks we run to get a collection? afair the full run of them takes 8 hours or so
> > > 
> > > 
> > > The benchmarks get split into 30 sub pieces and farmed out to helix, so I think we may run them all?
> > 
> > 
> > Yes, we run them all with `-iterationCount 1 --warmupCount 0 --invocationCount 1 --unrollFactor 1 ` so they complete quickly.
> 
> Meaning that they don't tier up properly?

Certainly seems possible, though with the aggressive tiering up that a checked runtime does, perhaps they do?

Let me first try and get this revamped so we are measuring a release runtime/SPC with a checked jit, and then we can see if we can afford to run the benchmarks more realistically.

**Comment by AndyAyersMS** [comment]
Another option is to just measure all release bits, but that probably makes the collection too brittle (?).

I suppose we could also enable the extra queries in release mode.

**Comment by BruceForstall** [comment]
> Let me first try and get this revamped so we are measuring a release runtime/SPC with a checked jit, and then we can see if we can afford to run the benchmarks more realistically.

Is it worth getting this checked in now, basically as-is, since it already is showing additional code variety, and trying to implement Release as a follow-up?

> Another option is to just measure all release bits, but that probably makes the collection too brittle (?).
> I suppose we could also enable the extra queries in release mode.

Enabling extra queries is Release is an interesting idea. It would be useful to do that, get collections, then ensure they can be replayed (with JitDisasm/JitDump).

**Comment by AndyAyersMS** [comment]
> > Let me first try and get this revamped so we are measuring a release runtime/SPC with a checked jit, and then we can see if we can afford to run the benchmarks more realistically.
> 
> Is it worth getting this checked in now, basically as-is, since it already is showing additional code variety, and trying to implement Release as a follow-up?

Not a bad idea, as having something now is probably better than having nothing...

**Comment by AndyAyersMS** [comment]
@BruceForstall addressed your feedback, so take another look when you can.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**